### PR TITLE
Extract solana-nonce crate

### PR DESCRIPTION
**Pull request by kevinheavey**
#### Problem
`solana_program::nonce` imposes a `solana_program` dependency on `solana_svm`

#### Summary of Changes
- Move to its own crate and re-export for backwards compatibility
- Make serde optional in the new crate

This branches off #2969 so that needs to be merged first

### CODE REVIEW

File: sdk/nonce/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+serde = [
+    "dep:serde",
+    "dep:serde_derive",
+    "solana-fee-calculator/serde",
+    "solana-hash/serde",
+    "solana-pubkey/serde",
+]
+
+[package.metadata.docs.rs]
+targets = ["x86_64-unknown-linux-gnu"]
```
Comment (joncinque) [inline sdk/nonce/Cargo.toml:35]: Wanna add the features bit and `#![cfg_attr(docsrs, feature(doc_auto_cfg))]`?

```suggestion
[package.metadata.docs.rs]
targets = ["x86_64-unknown-linux-gnu"]
all-features = true
rustdoc-args = ["--cfg=docsrs"]
```

**Review by joncinque — COMMENTED** [review]
The change looks good as is, while we're here, I'd like to do a little reorg for these files, since I always get lost in them. Could you:

* rename `state/mod.rs` to `versions.rs`, remove the re-export from current
* rename `state/current.rs` to `state.rs`

This makes some of the re-exports in solana-program a little gnarly, but I hope it'll be clearer at the base crate level. I believe we'll need to:

* re-export solana_nonce::state::State in nonce/mod.rs
* re-export `solana_nonce::versions::*` *and* `solana_nonce::state as current` in nonce/state/mod.rs
* remove nonce/state/current.rs

Or if it's too annoying, I can push those changes to your branch. Let me know!

File: sdk/program/src/nonce.rs [hunk 2]
```diff
@@ -0,0 +1,7 @@
+pub use solana_nonce::{state::State, NONCED_TX_MARKER_IX_INDEX};
+pub mod state {
+    pub use solana_nonce::{
+        state::{Data, DurableNonce, State},
```
Comment (joncinque) [inline sdk/program/src/nonce.rs:4]: Oh never mind, I had it wrong -- `current` isn't a public module, so there's no need to re-export it. Thanks!

**Review by joncinque — COMMENTED** [review]
Looks good to me! @yihau, please accept ownership of `solana-nonce` once it's passed

**Review by joncinque — APPROVED** [review]

**Comment by t-nelson** [comment]
is this not missing deprecation warnings for the old locations?

**Comment by joncinque** [comment]
> is this not missing deprecation warnings for the old locations?

Nice catch! Yeah we can add them. Since the deprecation warnings on re-exports don't actually do anything, it's not a big deal if they're missing. It's like forgetting comments.

**Comment by febo** [comment]
> > is this not missing deprecation warnings for the old locations?
> 
> Nice catch! Yeah we can add them. Since the deprecation warnings on re-exports don't actually do anything, it's not a big deal if they're missing. It's like forgetting comments.

It might be the case that clippy will actually start complaining about having them. 😅

https://github.com/rust-lang/rust/pull/132038

**Comment by t-nelson** [comment]
> > is this not missing deprecation warnings for the old locations?
> 
> Nice catch! Yeah we can add them. Since the deprecation warnings on re-exports don't actually do anything, it's not a big deal if they're missing. It's like forgetting comments.

yeah i've been thinking that we've been doing this backwards. we should be deprecating the symbols in their original location, then allowing deprecated on the new location/reexport. when we remove the deprecated code, that's when we actually move the symbols.

**Comment by kevinheavey** [comment]
> yeah i've been thinking that we've been doing this backwards. we should be deprecating the symbols in their original location, then allowing deprecated on the new location/reexport. when we remove the deprecated code, that's when we actually move the symbols.

Wouldn't that mean we'd be stuck with solana-program as a dependency everywhere until 3.0?

**Pull request merged**